components/ILIAS/OpenIdConnect/classes/class.ilOpenIdConnectSettingsGUI.php

```diff
@@ -672,4 +817,166 @@ protected function setSubTabs(string $active_tab): void
+    {
+        $this->showInfoMessage();
+        $this->setSubTabs(self::STAB_PROFILE);
+        if ((int) $this->mapping_template === 2) {
+            $this->userMapping();
+            return;
+        }
+
+        $this->initAttributeMapping();
+    }
+
+    private function showInfoMessage()
+    {
+        $url_text = $this->lng->txt('auth_oidc_here');
+
+        if ($this->mapping_template === 2) {
+            $url = $this->renderer->render($this->factory->link()->standard(
+                $this->lng->txt('auth_oidc_here'),
+                'https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims'
+            )
+                                                         ->withOpenInNewViewport(true));
+            $message = sprintf($this->lng->txt('auth_odic_scope_tab_info'), $url);
+        } else {
+            $url = $this->renderer->render($this->factory->link()->standard(
+                $this->lng->txt('auth_oidc_here'),
+                $this->ctrl->getLinkTarget($this, self::STAB_SCOPES)
+            ));
+            $tab_name = $this->lng->txt('auth_oidc_configured_scopes');
+            $message = sprintf($this->lng->txt('auth_odic_scope_info'), $url, $tab_name);
+        }
+
+        $this->renderer->render($this->factory->link()->standard(
+            $url_text,
+            $url
+        ));
+        $this->mainTemplate->setOnScreenMessage('info', $message);
+    }
+
+    private function initAttributeMapping(): void
+    {
+        $mapping = $this->attribute_mapping_template->getMappingRulesByAdditionalScopes($this->settings->getAdditionalScopes());
+        if (count($mapping) > 0) {
+            $this->settings->clearProfileMaps();
+        }
+        foreach ($mapping as $field => $item) {
+            $this->settings->setProfileMappingFieldValue(
+                $field,
+                $item
+            );
+        }
+
+        $this->userMapping();
+    }
+
+    /**
+     * @throws ilCtrlException
+     */
+    private function initUserMappingForm(): Standard
+    {
+        $ui_container = [];
+        $this->initUserDefinedFields();
+
+        foreach ($this->settings->getProfileMappingFields() as $mapping => $lang) {
+            $ui_container = $this->buildUserMappingInputForUserData($lang, $mapping, $ui_container);
+        }
+
+        foreach ($this->udf->getDefinitions() as $definition) {
+            $ui_container = $this->buildUserMappingInputFormUDF($definition, $ui_container);
+        }
+        $this->ctrl->setParameter(
+            $this,
+            "opic",
+            "opic_user_data_mapping"
+        );
+        return $this->ui->input()->container()->form()
+                        ->standard(
+                            $this->ctrl->getFormAction($this, 'saveProfile'),
+                            $ui_container
+                        );
+    }
+
+    protected function buildUserMappingInputFormUDF($definition, array $ui_container): array
```

Comment:
```suggestion
    private function buildUserMappingInputFormUDF($definition, array $ui_container): array
```
* The type of `$definition` is missing
* Please try to provide the type of `$ui_container` in the doc comments.
* Please try to provide the type of the return type in the doc comments.